libp2p/Session.cpp

```diff
@@ -305,8 +305,20 @@ void Session::disconnect(DisconnectReason _reason)
         RLPStream s;
         prep(s, DisconnectPacket, 1) << (int)_reason;
         sealAndSend(s);
+        auto disconnectTimer = m_server->createTimer();
+        auto self(shared_from_this());
+        disconnectTimer->expires_from_now(std::chrono::seconds(2));
+        disconnectTimer->async_wait([self, this, _reason](boost::system::error_code _ec) {
+            if(_ec)
+                LOG(m_netLoggerDetail)
+                    << "Error encountered while waiting after disconnect: " << _ec.message();
+            else
+                drop(_reason);
+        });
+        m_server->pollAsyncTimerHandler();
```

Comment:
What happens if you don't do this?

Reply:
@gumb0 
Generally we poll when we are calling the function to which the handler belongs multiple times (in a recursion or a loop) so that any handlers ready to run will start its execution.
The original motivation was when in a way I knew unit-tests were calling disconnect multiple times which later you confirmed that they create and destroy host multiple times.

But I realized that there are various parts of code that do call disconnect in a loop for eg.
https://github.com/ethereum/aleth/blob/4946ed615b11033db61ec47b2a97ef36bd3c047f/libp2p/Host.cpp#L247-L263
So I kept `pollAsyncTimerHandler` public in `Host` definition of which is here
https://github.com/twinstar26/aleth/blob/e83913941ace68e2b41597ddae7fa5ba8e6c3b6c/libp2p/Host.h#L266-L269

Reply:
Also I realize that there is a bug that we should be calling `_drop(DisconnectReason)` in here too right?
https://github.com/twinstar26/aleth/blob/e83913941ace68e2b41597ddae7fa5ba8e6c3b6c/libp2p/Session.cpp#L312-L314
So that if timer fails for some reason we will issue a drop. What do you think @gumb0 ?

Reply:
I actually don't think you need to call drop() there...I think that the handler will only ever called with a boost error code if the timer is cancelled but you never call cancel on the timer.